optapy-core/src/main/java/org/optaplanner/optapy/MirrorWithExtrasMap.java

```diff
@@ -93,4 +102,18 @@ public Set<Entry<Key_, Value_>> entrySet() {
         out.addAll(extraEntriesMap.entrySet());
         return out;
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof Map)) {
+            return false;
+        }
+        Map<?, ?> other = (Map) o;
+        return entrySet().equals(other.entrySet());
+    }
+
+    @Override
+    public int hashCode() {
+        return entrySet().hashCode();
+    }
```

Comment:
Why not just check equality/hashcode on the two maps?
In order to do it like this, you first have to create the entry set, and that will be needlessly wasteful.

Reply:
Also, while I'm looking at this code... why aren't these maps identity hash maps?
I'm wondering if key equality is really the behavior we want here. Would Python actually produce two different keys that are equal, and isn't this something we'd like to avoid?

Reply:
The keys in these maps are numbers; in particular, the address of the object (i.e. it is an IdentityHashMap). Because the same number can have different instances (i.e. there can be multiple instances of the number '2'), we need to use HashMap here, not IdentityHashMap

Reply:
`equals` need to go through the `entrySet`,as it is checking if combining `delegateMap` and `extraEntriesMap` results in the given `map`, which cannot be done with a single (or multiple) `equals` calls (without either creating the entry set or a new map). `hashCode` can be simplified though.

Reply:
I see. That said, this is some very heavy logic, and I hope `equals()` and `hashCode()` is not called on the hot path. Because it is, this will destroy any performance you have.

Reply:
It should not be called on the hot path (or at all; previously there was NOT an `equals`/`hashCode` implementation, which I noticed when testing it)

Reply:
`entrySet` should not be as expensive; I realize we already use `apache.collections`, which comes with a useful `CompositeSet`. Techically speaking, this class can probably be replaced with `CompositeMap` with a mutator strategy, but then we have nothing we can test since everything else goes to python, which lead to no `test` directory, which leads to unhappy sonar cloud.